File: 0787_cheapest_flight_within_k_stops/main.go
package main

import (
	"fmt"
	"math"
)

func buildGraph(flights [][]int) map[int]map[int]int {
	graph := map[int]map[int]int{}
	for _, flight := range flights {
		src, dst, cost := flight[0], flight[1], flight[2]
		cmap, ok := graph[src]
		if !ok {
			cmap = map[int]int{}
		}
		cmap[dst] = cost
		graph[src] = cmap
	}
	return graph
}

func findCheapestPrice(n int, flights [][]int, src int, dst int, K int) int {
	graph := buildGraph(flights)
	queue := [][]int{{src, 0}}
	stops := -1
	cheapest := math.MaxInt64
	visited := map[int]int{}
	for len(queue) > 0 && stops <= K {
		size := len(queue)
		for i := 0; i < size; i++ {
			curr := queue[i]
			from, sum := curr[0], curr[1]
			if from == dst && sum < cheapest {
				cheapest = sum
				continue
			}
			toMap := graph[from]
			for to, cost := range toMap {
				if min, ok := visited[to]; ok && sum+cost > min {
					continue
				}
				queue = append(queue, []int{to, sum + cost})
				visited[to] = sum + cost
			}
		}
		queue = queue[size:len(queue)]
		stops++
	}
	if cheapest == math.MaxInt64 {
		return -1
	}
	return cheapest
}

func main() {
	flights := [][]int{{0, 1, 100}, {1, 2, 100}, {0, 2, 500}}
	src := 0
	dst := 2
	K := 0
	fmt.Println(findCheapestPrice(3, flights, src, dst, K))
}

File: 0787_cheapest_flight_within_k_stops/main.py
from collections import deque


class Solution:
    def findCheapestPrice(self, n, flights, src, dst, K):
        """
        :type n: int
        :type flights: List[List[int]]
        :type src: int
        :type dst: int
        :type K: int
        :rtype: int
        """
        graph = self.buildGraph(flights)
        q = deque([(src, 0)])
        # stops start from -1 because K could be 0 (non-stop)
        stops = -1
        visited = {}
        cheapest = -1
        # we need to go to the dst to verify result, so <= K
        while len(q) > 0 and stops <= K:
            size = len(q)
            for i in range(size):
                start, total = q.popleft()
                if start == dst and (cheapest == -1 or cheapest > total):
                    cheapest = total
                    continue
                toMap = graph[start]
                for to, cost in toMap.items():
                    if to in visited and visited[to] > total + cost:
                        continue
                    q.append((to, total+cost))
                    visited[to] = total+cost
            stop += 1
        return cheapest

    def buildGraph(self, flights):
        graph = {}
        for flight in flights:
            src, dst, cost = flight
            if src in graph:
                graph[src] = {}
            cmap = graph[src]
            cmap[dst] = cost
            graph[src] = cmap
        return graph
                
            